### PR TITLE
Support GPU in contrainer

### PR DESCRIPTION
**Pull request by Sasasu**
example:
```xml
<?xml version="1.0" ?>
<configuration>
    <runtime>
        <id>plc_python_shared</id>
        <image>pivotaldata/plcontainer_python_shared:devel</image>
        <command>/clientdir/pyclient.sh</command>
        <shared_directory access="ro" container="/clientdir" host="/home/sa/GPDB/install/bin/plcontainer_clients"/>
        <setting enable_network="yes" />
        <device_request type="gpu" >
            <deviceid>0</deviceid>
            <deviceid>1</deviceid>
            <deviceid>2</deviceid>
            <deviceid>UUID=GPU-095f3011-889c-c65a-f103-e22637c28738</deviceid>
            <capacity>compute</capacity>
        </device_request>
    </runtime>
</configuration>
```

will transform to
```json5
{
  "DeviceRequests": [
    {
      "Count": 0, // docker require count = 0. when have DeviceIDs
      "DeviceIDs": [
        "0",
        "1",
        "2",
        "UUID=GPU-095f3011-889c-c65a-f103-e22637c28738"
      ],
      "Capabilities": [
        [
          "gpu",
          "compute"
        ]
      ]
    }
  ]
}
```

CUDA runtime is required inside the container. GTX graphics card is enough.
the host needs to install nvidia-docker, reference https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html

image

### CODE REVIEW

**Comment by Sasasu** [comment]
CI https://extensions.ci.gpdb.pivotal.io/teams/main/pipelines/DEV:SASAU:GPU:plcontainer-gpdb-6X-STABLE

File: src/plc_configuration.c [hunk 1]
```diff
@@ -330,6 +329,18 @@ static void parse_runtime_configuration(xmlNode *node) {
 					processed = 1;
 				}
 
+				if (xmlStrcmp(cur_node->name, (const xmlChar *) "device_request") == 0) {
```
Comment (RMTT) [inline src/plc_configuration.c:332]: The indentation seems wrong,
```
if (xmlStrcmp(cur_node->name, (const xmlChar *) "shared_directory") == 0)
```
should have same indentation with
```
if (xmlStrcmp(cur_node->name, (const xmlChar *) "device_request") == 0)
```

File: src/plc_configuration.c [hunk 2]
```diff
@@ -405,6 +416,82 @@ static void parse_runtime_configuration(xmlNode *node) {
...
+					value = NULL;
+				}
+
+				int c_i = 0, dev_i = 0;
+				for (xmlNode *sub = cur_node->children; sub != NULL; sub = sub->next) {
+					if (sub->type != XML_ELEMENT_NODE)
+						continue;
+
+#define XML_READ_TEXT(to, from) \
+					do { \
```
Comment (RMTT) [inline src/plc_configuration.c:466]: why using `do ... while` here?

**Comment by Sasasu** [comment]
sync with upstream. no logic change.

File: src/plc_configuration.c (hunk 1 above)
Comment (Sasasu) [reply src/plc_configuration.c:332]: clang-format it !

File: src/plc_configuration.c (hunk 2 above)
Comment (Sasasu) [reply src/plc_configuration.c:466]: used to simulate code block with ';' outside the macro.

```
if (xxx)
   XML_READ_TEXT();
else
   ;
```

will be
```
if (xxx)
   {xxxx}
   ;
else
   ;
```

https://stackoverflow.com/questions/1067226/c-multi-line-macro-do-while0-vs-scope-block

**Review by RMTT — APPROVED** [review]
```
demo=# select dummyPython();
                                   dummypython
---------------------------------------------------------------------------------
 Thu May 19 04:13:15 2022                                                       +
 +-----------------------------------------------------------------------------++
 | NVIDIA-SMI 515.43.04    Driver Version: 515.43.04    CUDA Version: 11.7     |+
 |-------------------------------+----------------------+----------------------++
 | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |+
 | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |+
 |                               |                      |               MIG M. |+
 |===============================+======================+======================|+
 |   0  Quadro T1000        Off  | 00000000:01:00.0 Off |                  N/A |+
 | N/A   48C    P8     3W /  N/A |      4MiB /  4096MiB |      0%      Default |+
 |                               |                      |                  N/A |+
 +-------------------------------+----------------------+----------------------++
                                                                                +
 +-----------------------------------------------------------------------------++
 | Processes:                                                                  |+
 |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |+
 |        ID   ID                                                   Usage      |+
 |=============================================================================|+
 +-----------------------------------------------------------------------------++

(1 row)
```
LGTM

**Comment by Sasasu** [comment]
Quadro T1000 is better than GTX 1050 :>

for the CI, there is no GPU environment for us to do such simple CUDA hello word.
I think a manual test is enough.

**Pull request merged**